### PR TITLE
Add Intel loop pipelining pass

### PR DESCRIPTION
**Pull request by chengjunlu**
This pass is like the upstream loop pipelining but for Intel specific. 
It optimize the loop body with tt.dot computation. 
Only work for the non-nested loop body at first.

### CODE REVIEW

**Review by etiotto — APPROVED** [review]

File: include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td [hunk 1]
```diff
@@ -101,6 +101,36 @@ def TritonIntelGPUDistributeToWarps
                            "mlir::gpu::GPUDialect"];
 }
 
+def TritonIntelGPUPipeline : Pass<"tritonintelgpu-pipeline", "mlir::ModuleOp"> {
+  let summary = "Pipeline loops";
+
+  let description = [{
+    Apply software pipelinining to loops containing expensive `tt.dot` operations.
```
Comment (whitneywhtsang) [inline include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td:108]: What kinds of tt.dot is considered as expensive, and what kinds are not?

File: include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td [hunk 2]
```diff
@@ -101,6 +101,36 @@ def TritonIntelGPUDistributeToWarps
...
+  let constructor = "mlir::triton::gpu::intel::createTritonIntelGPUPipelinePass()";
+
+  let dependentDialects = ["mlir::arith::ArithDialect",
+                           "mlir::scf::SCFDialect",
+                           "mlir::triton::TritonDialect",
+                           "mlir::triton::gpu::intel::TritonIntelGPUDialect"];
+
+  let options = [
+    Option<"numStages", "num-stages",
+           "int32_t", /*default*/"2",
```
Comment (whitneywhtsang) [inline include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td:122]: From experiments, prefetch distance of 3 is better for GEMM.

File: third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/SoftwarePipeliner.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+#include "mlir/Dialect/SCF/Transforms/Transforms.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonIntelGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonIntelGPU/Transforms/Passes.h"
+
+//===----------------------------------------------------------------------===//
+// This file will create a schedule that will be handed over to the pipeline
```
[the remark this answers is not in the document]
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/SoftwarePipeliner.cpp]: removed the comment

File: third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/MatmulLoopPipeline.cpp [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+  DenseSet<Operation *> prefetchAndDeps;
+  for (Operation *op : prefetchOps)
+    addDep(op, prefetchAndDeps, false);
+
+  // Find depenencies with distance of 1.
+  SmallVector<Operation *> distanceOneUsers;
+  for (Operation *op : prefetchAndDeps) {
+    for (Value operand : op->getOperands()) {
+#if 1
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/MatmulLoopPipeline.cpp:239]: remove

Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/MatmulLoopPipeline.cpp:239]: done

File: include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td (hunk 2 above)
Comment (etiotto) [reply include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td:122]: changed default

File: include/triton/Dialect/TritonIntelGPU/Transforms/Passes.h [hunk 5]
```diff
@@ -37,6 +37,12 @@ std::unique_ptr<Pass> createTritonIntelGPURewriteTensorPointerPass(
 std::unique_ptr<Pass> createPrefetchBlockPass();
 
 std::unique_ptr<Pass> createMatchTargetSizePass();
+
+std::unique_ptr<Pass>
+createTritonIntelGPUPipelinePass(int numStages = 2,
```
[the remark this answers is not in the document]
Comment (etiotto) [reply include/triton/Dialect/TritonIntelGPU/Transforms/Passes.h]: done

File: include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td [hunk 6]
```diff
@@ -246,4 +246,34 @@ def TritonIntelGPUMatchTargetSize : Pass<"tritonintelgpu-match-target-size", "ml
...
+  let constructor = "mlir::triton::gpu::intel::createTritonIntelGPUPipelinePass()";
+
+  let dependentDialects = ["mlir::triton::gpu::TritonGPUDialect",
+                           "mlir::triton::gpu::intel::TritonIntelGPUDialect",
+                           "mlir::scf::SCFDialect",
+                           "mlir::arith::ArithDialect"];
+
+  let options = [
+    Option<"numStages", "num-stages",
+           "int32_t", /*default*/"2",
```
[the remark this answers is not in the document]
Comment (etiotto) [reply include/triton/Dialect/TritonIntelGPU/Transforms/Passes.td]: done

File: third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/MatmulLoopPipeline.cpp [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+  return nullptr;
+}
+
+/// If all the transitive uses of the given value are used by a convert to the
+/// same dot operand encoding, return the encoding. Otherwise return nullptr.
+static ttg::DotOperandEncodingAttr allTransitiveUsesHaveDotEncoding(Value val) {
+  ttg::DotOperandEncodingAttr attr{nullptr};
+  for (Operation *user : val.getUsers()) {
+    ttg::DotOperandEncodingAttr dotAttr = getEncodingFromUser(user);
+    if (!dotAttr || (attr != nullptr && attr != dotAttr))
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUTransforms/Pipeliner/MatmulLoopPipeline.cpp:75]: use `all_of`?

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**